transports/quic/src/muxer.rs

```
@@ -89,6 +143,93 @@ impl QuicMuxer {
+        let mut inner = self.inner.lock();
+        inner.poll_connection(cx);
+        if let Some(substream_id) = inner.connection.pop_incoming_substream() {
+            inner.substreams.insert(substream_id, Default::default());
+            let substream = Substream::new(substream_id, self.inner.clone());
+            Poll::Ready(Ok(substream))
+        } else {
+            inner.poll_event_waker = Some(cx.waker().clone());
+            Poll::Pending
+        }
+    }
+
+    fn poll_outbound(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Self::Substream, Self::Error>> {
+        let mut inner = self.inner.lock();
+        inner.poll_connection(cx);
+        if let Some(substream_id) = inner.connection.pop_outgoing_substream() {
+            inner.substreams.insert(substream_id, Default::default());
+            let substream = Substream::new(substream_id, self.inner.clone());
+            Poll::Ready(Ok(substream))
+        } else {
+            inner.pending_substreams.push_back(cx.waker().clone());
+            Poll::Pending
+        }
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let mut inner = self.inner.lock();
+        inner.poll_connection(cx);
+
+        if inner.connection.connection.is_drained() {
+            return Poll::Ready(Ok(()));
+        }
+
+        if inner.substreams.is_empty() {
+            let connection = &mut inner.connection;
+            if !connection.connection.is_closed() {
+                connection.close();
+                if let Some(waker) = inner.poll_event_waker.take() {
+                    waker.wake();
+                }
+            } else {
+            }
+            while let Poll::Ready(event) = inner.connection.poll_event(cx) {
+                if let ConnectionEvent::ConnectionLost(_) = event {
+                    return Poll::Ready(Ok(()));
+                }
+            }
```

Let's assume this returns `Poll::Pending` at some point. Then we will break out of this loop and register a waker further down. Once we are woken, `poll_close` gets called again and we go straight into `inner.poll_connection` which will likely yield the `ConnectionEvent::ConnectionLost`. I don't think we will be observing it here then, right?

I think we may not want to call `poll_connection` at the top here or maybe return an error from `poll_connection` in case it has been closed?